engine/access/rpc/backend/backend_scripts.go

```diff
@@ -81,10 +84,45 @@ func (b *backendScripts) executeScriptOnExecutionNode(
 		Arguments: arguments,
 	}
 
-	execResp, err := b.executionRPC.ExecuteScriptAtBlockID(ctx, &execReq)
+	// if an executionRPC is provided, send the script to that execution node
+	if b.staticExecutionRPC != nil {
+		execResp, err := b.staticExecutionRPC.ExecuteScriptAtBlockID(ctx, &execReq)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to execute the script on the execution node: %v", err)
+		}
+		return execResp.GetValue(), nil
+	}
+
+	// find the execution nodes which have executed the block earlier and provided an execution receipt for it
+	execNodeIDs, err := executionNodesForBlockID(blockID, b.executionReceipts, b.state)
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "failed to execute the script on the execution node: %v", err)
+		return nil, err
+	}
+
+	var errors error
+	// choose upto 3 random execution node to try
+	execNodes := execNodeIDs.Sample(3)
+
+	// try to execute the script on one of the three execution nodes
+	for _, execNode := range execNodes {
+		result, err := b.tryExecuteScript(ctx, execNode, execReq)
+		if err == nil {
+			return result, nil
+		}
+		multierror.Append(errors, err)
 	}
+	return nil, errors
+}
 
+func (b *backendScripts) tryExecuteScript(ctx context.Context, execNode *flow.Identity, req execproto.ExecuteScriptAtBlockIDRequest) ([]byte, error) {
+	execRPCClient, closer, err := b.connFactory.GetExecutionAPIClient(execNode.Address)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to execute the script on the execution node %s: %v", execNode.String(), err)
+	}
+	defer closer.Close()
+	execResp, err := execRPCClient.ExecuteScriptAtBlockID(ctx, &req)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to execute the script on the execution node %s: %v", execNode.String(), err)
+	}
```

Comment:
Do we need to open and close these connections every time? Wouldn't it be more efficient to cache the client in the `ConnectionFactory` and re-use it?

Reply:
it will be more efficient to reuse the connection but its best handled by implementing a separate connection pooling in my opinion

Reply:
I guess the eventual goal is actually to replace `connFactory` with `connPool`